File: sal_v1.c
/*
SIMPLE ASSEMBLY LANGUAGE v1
winxos 2016-11-13
*/
#define _CRT_SECURE_NO_WARNINGS
//up for cancel visual studio 15' warnings (C4996: unsafe scanf).
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <io.h>

#define MAX_MEM 100
int MEM[MAX_MEM] = {0}, ADDER = 0; //ONLY MAX_MEN MEMORY, ONE REGISTER ADDER

enum
{
    INPUT = 10, PRINT, LOAD = 20, STORE, SET, ADD = 30, SUB, MUL, DIV, MOD,
    JMP = 40, JMPN, JMPZ, HALT = -1
};

int dir(char path[], char filter[])//list files in path with filter
{
    struct _finddata_t fi;
    long handle;
    if ((handle = _findfirst(strcat(strcat(path, "\\"), filter), &fi)) == -1L)
    {
        return -1;
    }
    do
    {
        if (!(fi.attrib & _A_SUBDIR))//not directory
        {
            printf("%s\n", fi.name);
        }
    } while (_findnext(handle, &fi) == 0);
    _findclose(handle);
    return 0;
}

int split(char dst[][80], char *str, const char *spl)//for single line chars split to words
{
    int n = 0;
    char *result = NULL;
    result = strtok(str, spl);
    while (result != NULL)
    {
        strcpy(dst[n++], result);
        result = strtok(NULL, spl);
    }
    return n;
}

int load_file(char file[80])//load .sml file to MEM
{
    FILE *fi;
    int i = 0;
    if (fi = fopen(file, "r"))
    {
        while (fscanf(fi, "%d", &MEM[i++]) != EOF);
    }
    else
    {
        printf("%s NOT FOUND.", file);
        return -1;
    }
    return 0;
}

void help()
{
    puts("SUPPORTED COMMAND:");
    puts("ls\t\t\t[list SML source codes]");
    puts("help | ? \t\t[help]");
    puts("load name.sml\t\t[load and run source code]");
    puts("input\t\t\t[open code input mode]");
    puts("exit\t\t\t[exit program]");
    puts("dump\t\t\t[show memories]");
    puts("\n");
}

void console()
{
    while (1)
    {
        printf("SML >");
        char buf[80] = {0};
        fgets(buf, 80, stdin);
        buf[strlen(buf) - 1] = 0;//remove \n
        char dst[10][80] = {0};
        int len = split(dst, buf, " ");
        if (0 == strcmp(dst[0], "ls"))
        {
            char path[2] = ".";
            if (-1 == dir(path, "*.sal"))
            {
                puts("NO .sal FILE FOUND ON CURRENT DIRECTORY.");
            }
        }
        else if (0 == strcmp(dst[0], "help") || 0 == strcmp(dst[0], "?"))
        {
            help();
        }
        else if (0 == strcmp(dst[0], "exit"))
        {
            puts("See you soon.");
            system("ping -n 3 127.1>nul");
            exit(0);
        }
        fflush(stdin);
        rewind(stdin); //
    }
}

int
main()
{
    printf("\tSimple Assembly Language Simulator v1.0\n");
    printf("\t\t\twinxos 2016/11/13\n\n\n");
    help();
    console();
    return 0;
}

File: sal_v1.py
# coding:utf-8
# SIMPLE ASSEMBLY LANGUAGE V1
# ASSEMBLY SAL FILE TO SAL FOR SML
# PYTHON 3.5
# winxos 2016-11-15
import glob
import sys, os

keywords = {"INPUT": 10, "PRINT": 11,
            "LOAD": 20, "STORE": 21, "SET": 22,
            "ADD": 30, "SUB": 31, "MUL": 32, "DIV": 33, "MOD": 34, "INC": 35, "DEC": 36, "NEG": 37,
            "JMP": 40, "JMPN": 41, "JMPZ": 42, "HALT": 43,
            "AND": 50, "OR": 51, "XOR": 52,
            "PUSH": 60, "POP": 61
            };
MAX_MEM = 100
var_table = {}
label_table = {}


def pretranslate(src_code):
    after_code = []
    for line in src_code:
        cmds = str(line).split()
        if len(cmds) > 0:
            if cmds[0] == "DIM":  # vars
                if len(cmds) == 2:
                    cmds.append("0")
                if len(cmds) > 2:
                    if cmds[1] in var_table:
                        print("[error] VAR %s ALREADY DEFINED." % cmds[1])
                        return
                    var_table[cmds[1]] = cmds[2]
            elif cmds[0].startswith("."):  # label
                if cmds[0] in label_table:
                    print("[error] LABEL %s ALREADY DEFINED." % cmds[0])
                    return
                label_table[cmds[0]] = len(after_code)
            elif cmds[0].startswith(";"):  # comment
                pass
            else:  # normal code
                after_code.append(cmds)

    for i, n in enumerate(var_table):
        after_code.append(var_table[n])
        var_table[n] = len(after_code) - 1
    return after_code


def translate(codes):
    ml = []
    for index, code in enumerate(codes):
        if index >= len(codes) - len(var_table):
            ml += (codes[-len(var_table):])
            break
        operator = keywords[code[0]]
        operand = 0
        if code[0] == "JMP" or code[0] == "JMPN" or code[0] == "JMPZ":
            if len(code) > 1:
                label = "." + code[1]
                if label in label_table:
                    operand = label_table[label]
                else:
                    print("[error] LABEL %S NOT FOUND." % code[1])
                    return;
            else:
                print("[waring] LINE % MISS LABEL." % index)
        else:
            if len(code) > 1:
                if code[1] in var_table:
                    operand = var_table[code[1]]
                else:
                    print("[error] VAR %s NOT FOUND." % code[1])
                    return;
            else:
                if code[0] == "HALT" or code[0] == "PUSH" or code[0] == "POP":
                    pass
                else:
                    print("[waring] LINE % MISS OPERAND." % index)
        ml.append("%02d%02d" % (operator, operand))
    return ("\n").join(ml)


def deal_sal(n):
    fn = str(n).split(".")[0] + ".sml"
    try:
        with open(n) as fi, open(fn, "w") as fo:
            fo.write(translate(pretranslate(fi.read().split("\n"))))
            print("[success] ASSEMBLY FILE %s TO %s" % (n, fn))
    except IOError as e:
        print("[error] %s" % str(e))


def console():
    print("WELCOME TO SIMPLE ASSEMBLER v1")
    print("\t\twinxos 2016-11-15")
    print("\n")
    print("ls\t\t[list .sal assembly file]")
    print("as file\t\t[assembly .sal file to .sml file]")
    print("exit\t\t[exit program]")
    while True:
        cmd = str(input("SAL >"))
        cmds = cmd.split(" ")
        if cmds[0] == "ls":
            print("\n".join(glob.glob(r"*.sal")))
        if cmds[0] == "as":
            deal_sal(cmds[1])
        if cmds[0] == "exit":
            sys.exit()


if __name__ == "__main__":
    if len(sys.argv) > 1:
        deal_sal(sys.argv[1])
    else:
        console()
